hci.c

```diff
@@ -0,0 +1,217 @@
+    return e;
+}
+
+static
+void hci_update_query_timer(HCIManager * manager)
+{
+    double e = hci_now(manager);
+    e = e - manager->timer_query_begin;
+    if(e > manager->LongestTimeBetweenQueries)
+        manager->LongestTimeBetweenQueries = e;
+
+    manager->timer_query_begin = e;
+}
+
+/*
+ * query the filesystem for HCI commands;
+ * returns the content of the file or NULL; collectively
+ * */
+static char *
+hci_query_filesystem(HCIManager * manager, char * filename)
+{
+    int ThisTask;
+    int NTask;
+    MPI_Comm_size(MPI_COMM_WORLD, &NTask);
+    MPI_Comm_rank(MPI_COMM_WORLD, &ThisTask);
+    int size = 0;
+    char * content = NULL;
+    if(ThisTask == 0) {
+        char * fullname = fastpm_strdup_printf("%s/%s", manager->prefix, filename);
+        content = fastpm_file_get_content(fullname);
+        if(content) {
+            size = strlen(content);
+            remove(fullname);
+        } else {
+            size = -1;
+        }
+        free(fullname);
+    }
+    MPI_Bcast(&size, 1, MPI_INT, 0, MPI_COMM_WORLD);
+
+    if(size != -1) {
+        if(ThisTask != 0) {
+            content = calloc(size + 1, 1);
+        }
+        MPI_Bcast(content, size+1, MPI_BYTE, 0, MPI_COMM_WORLD);
+    } else {
+        content == NULL;
+    }
+
+    return content;
+}
+
+static char *
+hci_query_timeout(HCIManager * manager)
+{
+    double now = hci_get_elapsed_time(manager);
+    /*
+     * factor 0.9 is a safety tolerance
+     * for possible inconsistency between measured time and the true wallclock
+     *
+     * If there likely isn't time for a new query, then we shall timeout as well.
+     * */
+
+    if (now + manager->LongestTimeBetweenQueries < manager->WallClockTimeLimit * 0.9) {
+        return NULL;
+    }
+
+    /* any freeable string would work. */
+    return calloc(32, 1);
+}
+
+static char *
+hci_query_auto_checkpoint(HCIManager * manager)
+{
+    /*How long since the last checkpoint?*/
+    if(manager->AutoCheckPointTime <= 0) return NULL;
+
+    double now = hci_get_elapsed_time(manager);
+    if(now - manager->TimeLastCheckPoint >= manager->AutoCheckPointTime) {
+        /* any freeable string would work */
+        return calloc(32, 1);
+    }
+    return NULL;
+}
+
+/*
+ * the return value is non-zero if the mainloop shall break.
+ * */
+int
+hci_query(HCIManager * manager, HCIAction * action)
+{
+    /* measure time since last query */
+    hci_update_query_timer(manager);
+
+    /* Check whether we need to interrupt the run */
+
+    char * request;
+
+    /* Will we run out of time by the query ? highest priority.
+     */
+    if(request = hci_query_timeout(manager)) {
+        message(0, "Stopping due to TimeLimitCPU, dumping a CheckPoint.\n");
+        action->type = HCI_TIMEOUT;
+        action->write_snapshot = 1;
+        free(request);
+        return 1;
+    }
+
+    if(request = hci_query_filesystem(manager, "ioctl"))
+    {
+        //update_IO_params(request);
```

Comment:
This should not be commented!

Reply:
I disabled it because I think this needs to be replaced by something more general, e.g. rerunning the configuration parser, and copy over the useful new parameters.

Reply:
I think that would be a different pull request. I don't use this feature personally, but it seems strange to disable it by a commented line without replacement. Adding a fixme comment would be better.

Reply:
Yes. It shall be a fixme. I added a warning too. I don't think anyone actually use this feature anymore.